From 6ba132a427722851805b6792af7e55c25e468754 Mon Sep 17 00:00:00 2001
From: Sam Edwards <CFSworks@gmail.com>
Date: Fri, 27 Oct 2023 15:34:28 -0700
Subject: [PATCH 17/17] board: turing: pi2: add board definition

This is very hacky and not really upstream-worthy. I am hoping that
the code can still be useful, but this needs its own Makefile/Kconfig,
which is difficult since sunxi doesn't seem friendly to board/ dirs??
---
 arch/arm/mach-sunxi/Kconfig      |  4 ++++
 arch/arm/mach-sunxi/board.c      |  4 ++++
 board/sunxi/Makefile             |  2 ++
 board/turing/pi2/turingpi2-spl.c | 29 +++++++++++++++++++++++++++++
 4 files changed, 39 insertions(+)
 create mode 100644 board/turing/pi2/turingpi2-spl.c

diff --git a/arch/arm/mach-sunxi/Kconfig b/arch/arm/mach-sunxi/Kconfig
index 1924967f14..011e46ccff 100644
--- a/arch/arm/mach-sunxi/Kconfig
+++ b/arch/arm/mach-sunxi/Kconfig
@@ -1111,6 +1111,10 @@ config BLUETOOTH_DT_DEVICE_FIXUP
 	  The used address is "bdaddr" if set, and "ethaddr" with the LSB
 	  flipped elsewise.
 
+config TARGET_TURINGPI2
+	bool "Turing Pi 2"
+	depends on MACH_SUN8I_R528
+
 source "board/sunxi/Kconfig"
 
 endif
diff --git a/arch/arm/mach-sunxi/board.c b/arch/arm/mach-sunxi/board.c
index 829d170aeb..59f58f77e9 100644
--- a/arch/arm/mach-sunxi/board.c
+++ b/arch/arm/mach-sunxi/board.c
@@ -481,6 +481,10 @@ void board_init_f(ulong dummy)
 	tzpc_init();
 #endif
 
+#if defined(CONFIG_BOARD_EARLY_INIT_F)
+	board_early_init_f();
+#endif
+
 	clock_init();
 	timer_init();
 	gpio_init();
diff --git a/board/sunxi/Makefile b/board/sunxi/Makefile
index d96b7897b6..0ee947bac0 100644
--- a/board/sunxi/Makefile
+++ b/board/sunxi/Makefile
@@ -12,3 +12,5 @@ obj-$(CONFIG_MACH_SUN4I)	+= dram_sun4i_auto.o
 obj-$(CONFIG_MACH_SUN5I)	+= dram_sun5i_auto.o
 obj-$(CONFIG_MACH_SUN7I)	+= dram_sun5i_auto.o
 obj-$(CONFIG_CHIP_DIP_SCAN)	+= chip.o
+
+obj-$(CONFIG_TARGET_TURINGPI2)	+= ../turing/pi2/turingpi2-spl.o
diff --git a/board/turing/pi2/turingpi2-spl.c b/board/turing/pi2/turingpi2-spl.c
new file mode 100644
index 0000000000..6df32f53d4
--- /dev/null
+++ b/board/turing/pi2/turingpi2-spl.c
@@ -0,0 +1,29 @@
+// SPDX-License-Identifier: GPL-2.0+
+/*
+ * Copyright (C) 2023 Sam Edwards <CFSworks@gmail.com>
+ *
+ * Early init for the Turing Pi 2 clusterboard.
+ */
+
+#include <common.h>
+#include <asm/gpio.h>
+#include <init.h>
+#include <sunxi_gpio.h>
+
+static void turingpi2_ethsw_rst(void)
+{
+	/*
+	 * RTL8370MB switch reset (active-low) on PG13 should be asserted very
+	 * early in boot to prevent Ethernet from coming up until the switch
+	 * can be configured. The user may be using the Turing Pi 2 as a router
+	 * or some other kind of network isolation device.
+	 */
+	gpio_direction_output(SUNXI_GPG(13), 0);
+}
+
+int board_early_init_f(void)
+{
+	turingpi2_ethsw_rst();
+
+	return 0;
+}
-- 
2.44.0

